### PR TITLE
[Vaults UI] Add groupIds to conversation

### PR DESCRIPTION
**Pull request by philipperolet**
## Description
Preliminary to PR #7631 and to fix issue #6903, following the decision to denormalize groups on conversations

This PR creates a `groupIds` field on conversations based on the ACL of messages it contains. The field will be populated / updated:
- at postUserMessage;
- at editUserMessage;
- at retryAgentMessage;

### Cases with special attention

**at editUserMessage--if a mention is removed** and for when  we implement the ability to delete an agent message from a conversation

Currently, it does not delete former agent messages if mention is changed so the convo should not loosen its permission.

In the future, there's a question of updating convo permission if a message is deleted. But we'd have to be careful in depending on how we implement it, e.g. if we still keep the agent messages versions when deleted just as we keep the various messages versions, then the best would be to  keep a "purely additive" stance and *not* remove permission restriction

The purely additive model seems cleaner, yielding simpler code,  and user downside very limited (only case is: now that the message is deleted user should be able to access the convo but can't, seems very edge-casey and easily overcomable)

This is why no step is taken in this PR towards permission loosening on message deletion.

**when agent groupIds change**

We do not update conversations' groupIds at those points, see [thread](https://dust4ai.slack.com/archives/C050SM8NSPK/p1727423063794339). Mainly because the update will happen naturally next time the agent is pinged, and older conversation messages are legitimate to keep the former permissioning for a while.

**when a vault's groups change**
The denormalization also raises the question of vaults whose groups would change since theoretically, vaults <-> groups is a many-to-many relationship. Updating a vault's groups should trigger an update of all agentmessages.

In practice though, it is currently a one-to-one relationship and there is no code location in which the relationship can mutate after creation. This case is therefore *not* adressed.

## Risk
Field is not used yet, so risk is limited.
Tests: 
- check conversations are properly backfilled
- check posting a message with a restricted agent fills convo groupid properly
- check retrying a message with a newly restricted agent does the same

## Deploy Plan
- run migration
- deploy front
- run backfill

### CODE REVIEW

File: front/lib/api/assistant/conversation.ts [hunk 1]
```diff
@@ -1336,6 +1352,13 @@ export async function* editUserMessage(
         row: AgentMessage;
         m: AgentMessageWithRankType;
       }[];
+
```
Comment (philipperolet) [inline front/lib/api/assistant/conversation.ts:1355]: I revised my stance on editUserMessage => currently, it does not delete former agent messages if mention is changed so the convo should not loosen its permission.

We'd have to be careful in the future when deletion comes and depending on how we implement it, e.g. if we still keep the agent messages versions when deleted just as we keep the various messages versions, then we'd actually keep the "purely additive" stance and *not* remove permission restriction

Thinking about it more, it seems better: simpler code, clearer additive model, and user downside very limited (only case is: now that the message is deleted user should be able to access the convo but can't, seems very edge-casey and easily overcomable)

Comment (philipperolet) [reply front/lib/api/assistant/conversation.ts:1355]: Will edit the PR description if there's agreement

File: front/lib/api/assistant/participants.ts [hunk 2]
```diff
@@ -69,7 +69,6 @@ export async function fetchConversationParticipants(
     where: {
       conversationId: conversation.id,
     },
-    order: [["rank", "ASC"]],
```
Comment (spolu) [inline front/lib/api/assistant/participants.ts:72]: Why change that?

File: front/lib/api/assistant/conversation.ts (hunk 1 above)
Comment (spolu) [reply front/lib/api/assistant/conversation.ts:1355]: Let's leave a comment here?

File: front/lib/models/assistant/conversation.ts [hunk 3]
```diff
@@ -32,6 +32,8 @@ export class Conversation extends Model<
   declare title: string | null;
   declare visibility: CreationOptional<ConversationVisibility>;
 
+  declare groupIds: number[];
```
Comment (spolu) [inline front/lib/models/assistant/conversation.ts:35]: Is it possible to type it more as a foreign key?

File: front/migrations/20240927_backfill_conversations_groupIds.ts [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  }
+}
+
+async function updateConversation(
+  auth: Authenticator,
+  conversation: Conversation,
+  execute: boolean,
+  logger: Logger
+) {
+  const messages = await Message.findAll({
```
Comment (spolu) [inline front/migrations/20240927_backfill_conversations_groupIds.ts:88]: Probably worth a comment wrt to why we don't need to get last by rank here?

**Review by spolu — APPROVED** [review]
LGTM left 2 comments and a question

File: front/lib/api/assistant/participants.ts (hunk 2 above)
Comment (philipperolet) [reply front/lib/api/assistant/participants.ts:72]: At @flvndvd request, and I agree with him, to speed up the call. The ordering is not needed and prevents the index from working efficiently (admittedly out of scope of PR but well)

File: front/lib/models/assistant/conversation.ts (hunk 3 above)
Comment (philipperolet) [reply front/lib/models/assistant/conversation.ts:35]: we would need to have a join table, which I think is overkill. We can discuss and change it later if you think it's worth it

Comment (spolu) [reply front/lib/models/assistant/conversation.ts:35]: Agreed it's not worth it

**Pull request merged**